### PR TITLE
rtc: bump ffi to v0.10.2

### PR DESCRIPTION
**Pull request by nbsp**

### CODE REVIEW

**Review by theomonnom — APPROVED** [review]
Nice, can you add the new parameters inside the AudioSource. (buffer_size_ms)
There is also:
wait_for_playout
clear_queue

**Comment by nbsp** [comment]
@theomonnom can you re-check? want to make sure i did this impl correctly

File: packages/livekit-rtc/src/audio_source.ts [hunk 1]
```diff
@@ -47,7 +63,46 @@ export class AudioSource {
     this.ffiHandle = new FfiHandle(res.source.handle.id);
   }
 
+  get queuedDuration(): number {
+    return Math.max(this.currentQueueSize - Date.now() + this.lastCapture, 0);
```
Comment (theomonnom) [inline packages/livekit-rtc/src/audio_source.ts:67]: Date.now isn't monotonic, can you use https://nodejs.org/api/process.html#process_process_hrtime_bigint

File: packages/livekit-rtc/src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+// SPDX-FileCopyrightText: 2024 LiveKit, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+import EventEmitter, { once } from 'events';
+
+/** @internal */
+export class Queue<T> {
```
Comment (theomonnom) [inline packages/livekit-rtc/src/utils.ts:7]: Seems like the Queue isn't needed, just Promise/Future?

Comment (nbsp) [reply packages/livekit-rtc/src/utils.ts:7]: this promise is reused more than one time so it's more ergonomic to use queue for this

File: packages/livekit-rtc/src/audio_source.ts [hunk 3]
```diff
@@ -47,7 +63,49 @@ export class AudioSource {
...
   async captureFrame(frame: AudioFrame) {
+    const now = Number(process.hrtime.bigint() / 1000000n);
+    const elapsed = this.lastCapture === 0 ? 0 : now - this.lastCapture;
+    this.currentQueueSize += (frame.samplesPerChannel / frame.sampleRate - elapsed) * 1000;
+
+    // remove 50ms to account for processing time (e.g. using wait_for_playout for very small chunks)
+    this.currentQueueSize -= 50;
+    this.lastCapture = now;
+
+    if (this.timeout) {
```
Comment (theomonnom) [inline packages/livekit-rtc/src/audio_source.ts:104]: Seems like this.timeout is never set?

Is `waitForPlayout` working correctly?

Comment (nbsp) [reply packages/livekit-rtc/src/audio_source.ts:104]: oops yeah. it seemed to work fine enough so i didn't notice.

File: packages/livekit-rtc/src/audio_source.ts [hunk 4]
```diff
@@ -47,7 +63,50 @@ export class AudioSource {
...
+    });
+
+    FfiClient.instance.request<ClearAudioBufferResponse>({
+      message: {
+        case: 'clearAudioBuffer',
+        value: req,
+      },
+    });
+
+    this.releaseQueue.put();
```
Comment (theomonnom) [inline packages/livekit-rtc/src/audio_source.ts:85]: if clearQueue is called multiple times, it seems like waitForPlayout will directly returns?

**Review by theomonnom — CHANGES_REQUESTED** [review]

File: packages/livekit-rtc/src/audio_source.ts (hunk 4 above)
Comment (nbsp) [reply packages/livekit-rtc/src/audio_source.ts:85]: right this isn't one-to-one. i'll rework this to use Futures

**Review by davidzhao — APPROVED** [review]

**Review by theomonnom — APPROVED** [review]

**Pull request merged**